src/main/java/com/google/sps/servlets/BlobServlet.java

```diff
@@ -33,59 +41,203 @@
   public void doPost(HttpServletRequest request, HttpServletResponse response)
       throws IOException {
     UserService userService = UserServiceFactory.getUserService();
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
 
+    // Must be logged in
     if (!userService.isUserLoggedIn()) {
-      response.sendRedirect("/imgupload.html");
+      throw new IOException("User must be logged in.");
```

Comment:
what catches this exception?  what will the user see if this happens? is this never supposed to happen?  If this is user facing then another option would be to redirect them to the login page.

Instead of asking this every time you throw an exception I am only going to ask it here.  If there are changes to be made then I would do it for every exception.